helper/funcs_test.go

```diff
@@ -245,70 +246,104 @@ func TestCheckNamespaceScope(t *testing.T) {
 		},
 		{
-			name:     "ok relative path inside sandbox",
+			name:     "fail unjoined absolute path outside sandbox",
+			path:     "/secrets",
+			dir:      "/alloc",
+			join:     false,
+			expected: true,
+		},
+		{
+			name:     "ok joined relative path inside sandbox",
+			path:     "./safe",
+			dir:      "/alloc",
+			join:     true,
+			expected: false,
+		},
+		{
+			name:     "fail unjoined relative path outside sandbox",
 			path:     "./safe",
 			dir:      "/alloc",
-			expected: "/alloc/safe",
+			join:     false,
+			expected: true,
 		},
 		{
 			name:     "ok relative path traversal constrained to sandbox",
 			path:     "../../alloc/safe",
 			dir:      "/alloc",
-			expected: "/alloc/safe",
+			join:     true,
+			expected: false,
 		},
 		{
-			name:     "ok absolute path traversal constrained to sandbox",
+			name:     "ok unjoined absolute path traversal constrained to sandbox",
 			path:     "/../alloc/safe",
 			dir:      "/alloc",
-			expected: "/alloc/safe",
+			join:     true,
+			expected: false,
 		},
 		{
-			name:        "fail absolute path outside sandbox",
-			path:        "/unsafe",
-			dir:         "/alloc",
-			expected:    "/unsafe",
-			expectedErr: "\"/unsafe\" escapes sandbox directory",
+			name:     "ok unjoined absolute path traversal constrained to sandbox",
+			path:     "/../alloc/safe",
+			dir:      "/alloc",
+			join:     false,
+			expected: false,
 		},
 		{
-			name:        "fail relative path traverses outside sandbox",
-			path:        "../../../unsafe",
-			dir:         "/alloc",
-			expected:    "/unsafe",
-			expectedErr: "\"/unsafe\" escapes sandbox directory",
+			name:     "fail joined relative path traverses outside sandbox",
+			path:     "../../../unsafe",
+			dir:      "/alloc",
+			join:     true,
+			expected: true,
 		},
 		{
-			name:        "fail absolute path tries to transverse outside sandbox",
-			path:        "/alloc/../unsafe",
-			dir:         "/alloc",
-			expected:    "/unsafe",
-			expectedErr: "\"/unsafe\" escapes sandbox directory",
+			name:     "fail unjoined relative path traverses outside sandbox",
+			path:     "../../../unsafe",
+			dir:      "/alloc",
+			join:     false,
+			expected: true,
+		},
+		{
+			name:     "fail joined absolute path tries to transverse outside sandbox",
+			path:     "/alloc/../../unsafe",
+			dir:      "/alloc",
+			join:     true,
+			expected: true,
+		},
+		{
+			name:     "fail unjoined absolute path tries to transverse outside sandbox",
+			path:     "/alloc/../../unsafe",
+			dir:      "/alloc",
+			join:     false,
+			expected: true,
 		},
 	}
 
 	for _, tc := range cases {
 		t.Run(tc.name, func(t *testing.T) {
 			caseMsg := fmt.Sprintf("path: %v\ndir: %v", tc.path, tc.dir)
-			escapes, err := GetPathInSandbox(tc.dir, tc.path)
-			if tc.expectedErr != "" {
-				require.EqualError(t, err, tc.expectedErr, caseMsg)
-			} else {
-				require.NoError(t, err, caseMsg)
+
+			path := tc.path
+			if tc.join {
```

Comment:
I'm finding the `if tc.join` here makes it hard to follow, because the caller to `PathEscapesSandbox` (with this change) _always_ calls `filepath.Join` on `config.TaskDir` and `src`. It might be clearer in the tests if the `path` included the `dir`, with the `tc.join` stuff removed entirely.

Reply:
I have this here because there are 3 callers: the two that unconditionally join the paths and the one in template that only joins if the `source` is absolute ([ref](https://github.com/hashicorp/nomad/pull/9149/files#diff-796a6e8da6dbce4e38ef692cf9f9e70b5b5770f9cf367e3e13ca1f86de356d01R556))

Reply:
Technically whether the path is joined doesn't matter to the `PathEscapesSandbox` at all, so maybe it'd be better just to have the paths "pre-joined" for clarity where appropriate?

Reply:
The pre-joining is what I was looking for; it seems to exist on some tests already (the original ones?).

Reply:
Done in 2505858